Mathlib/Analysis/Convex/Basic.lean

```diff
@@ -480,11 +480,11 @@ theorem Convex_subadditive_le [SMul 𝕜 E] {f : E → 𝕜} (hf1 : ∀ x y, f (
     Convex 𝕜 { x | f x ≤ B } := by
   rw [convex_iff_segment_subset]
   rintro x hx y hy z ⟨a, b, ha, hb, hs, rfl⟩
-  calc
-    _ ≤ a • (f x) + b • (f y) := le_trans (hf1 _ _) (add_le_add (hf2 x ha) (hf2 y hb))
-    _ ≤ a • B + b • B :=
-        add_le_add (smul_le_smul_of_nonneg_left hx ha) (smul_le_smul_of_nonneg_left hy hb)
-    _ ≤ B := by rw [← add_smul, hs, one_smul]
+  dsimp at hx hy ⊢
+  linear_combination (norm := skip)
+    hf1 (a • x) (b • y) + hf2 x ha + hf2 y hb + a * hx + b * hy + hs * B
+  apply le_of_eq
+  noncomm_ring [smul_eq_mul]
```

Comment:
This is the only remaining change where I think we're substantially loosing some readability.
Is it possible to keep the `calc` block but change its subproofs?

Reply:
Yeah, I agree.  The basic problem here is that there's no particularly good discharger tactic for noncommutative-ordered-rings (!).  It was a good exercise to try to make `linear_combination` work in this idiom, but I conclude that it's not worth the effort.  Reverted.